### PR TITLE
Added segmented fill for hpx::vector

### PR DESCRIPTION
**Pull request by Sanac**
Adding segmented fill for `hpx::vector`. (refer to [#1338](https://github.com/STEllAR-GROUP/hpx/issues/1338))

### CODE REVIEW

**Comment by hkaiser** [comment]
This is great! 

Could this be somehow merged with the segmented `for_each`? While it might be necessary, it would be nice if we could avoid reimplementing each of the algorithms separately.

**Comment by Sanac** [comment]
Rewritten so that `segmented_fill` uses `for_each`.

**Comment by biddisco** [comment]
I've not yet been through the code completely, but we discussed (via skype), that the tests should be extended to run distributed as well as locally. This will require some cmake additions that we'll put on another PR once they are done.

**Comment by hkaiser** [comment]
@biddisco What changes to the build/test system do you have in mind?

**Comment by biddisco** [comment]
Only some extra cmake code in `tests/unit/component/CMakeLists.txt ` to add tests for partition_vector algorithms into distributed tests as well.

I believe that any partitioned_vector tests should be run with distributed as well as local partitions and I was not able to locate any existing tests of them like this.

**Comment by hkaiser** [comment]
> I believe that any partitioned_vector tests should be run with distributed as well as local partitions and I was not able to locate any existing tests of them like this.

I agree. I'm just not sure what those tests would need in addition to the other distributed tests we already run.

File: hpx/parallel/segmented_algorithms/fill.hpp [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+            return result::get(
+                dataflow(
+                    [=](decltype(r) && res)
+                    {
+                        hpx::util::unwrapped(res);
+                        return;
+                    },
+                    std::move(r)));
+        }
```
Comment (hkaiser) [inline hpx/parallel/segmented_algorithms/fill.hpp]: @Sanac What is the rationale of separating the synchronous and asynchronous versions here. Wouldn't both be equivalent, letting the underlying `parallel::for_each` handle the differences?:

         template <typename Algo, typename ExPolicy, typename SegIter, typename T>  
         static typename util::detail::algorithm_result<ExPolicy, void>::type  
         fill_(ExPolicy && policy, SegIter first, SegIter last, T const& value)  
         {  
             return hpx::parallel::for_each(std::forward<ExPolicy>(policy),
                  first, last, fill_function<T>(value));  
        }

Comment (hkaiser) [inline hpx/parallel/segmented_algorithms/fill.hpp]: @Sanac: well, it probably should be:

         template <typename Algo, typename ExPolicy, typename SegIter, typename T>  
         static typename util::detail::algorithm_result<ExPolicy, void>::type  
         fill_(ExPolicy && policy, SegIter first, SegIter last, T const& value)  
         {  
             typedef typename util::detail::algorithm_result<ExPolicy, void>::type  
                 result_type;
             return hpx::util::void_guard<result_type>(),
                 hpx::parallel::for_each(std::forward<ExPolicy>(policy),
                     first, last, fill_function<T>(value));
         }

**Comment by hkaiser** [comment]
@Sanac thanks for those changes! One last question: what's the rationale of making the implementation functions for the new algorithm non-static?

**Comment by hkaiser** [comment]
@Sanac I take that back, you changed from `inline` to `static`, not v.v. 
Everything looks good to me now, thanks!

**Comment by hkaiser** [comment]
@Sanac once the inspect errors are fixed (see https://circle-artifacts.com/gh/STEllAR-GROUP/hpx/4174/artifacts/0/tmp/circle-artifacts.ACeOWQE/hpx_inspect_report.html) we can go ahead and merge your changes.

**Comment by hkaiser** [comment]
Thanks!

**Pull request merged**